### ISSUE TITLE
Compilation failed after re2 update

### ISSUE THREAD

**Issue by netDalek** [tuncer/re2#10]
First of all thank you for this library.

After last re2 update I had to add https://github.com/netDalek/re2/commit/f2a2e8a70c84d0183891251a8826eec008e5f253 to make script. Because compilation failed with
```
Cloning into 're2'...
c++ -m64 -c -o obj/util/hash.o  -Wall -O3 -fPIC -pthread -m64 -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -I.     -DNDEBUG util/hash.cc
c++ -m64 -c -o obj/util/logging.o  -Wall -O3 -fPIC -pthread -m64 -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -I.     -DNDEBUG util/logging.cc
In file included from util/hash.cc:41In file included from ./util/logging.h:13,
                 from util/logging.cc:5:
./util/util.h:33:18: error: atomic: No such file or directory
:
./util/util.h:33:18: error: atomic: No such file or directory
```

And probably it would be better to specify exact https://code.googlesource.com/re2 version to avoid such errors

**Comment by tuncer** [comment]
I'm afraid newer re2 upstream requires (#6) C++11 to build, and there is nothing we can do about it. On platforms without a C++11 compiler you can checkout tuncer/re2 v2.2.2 (bb5adb417f59fca4a2463fbc0aabd1f1f9f9b347), where I've pinned it to upstream tag 2006-02-01. Because there is no C++98 branch to follow upstream, I cannot choose to avoid it. So, unless the C++11-dependency is removed upstream, the next upstream release will need a newer C++ toolchain.

What platform is this and how common of an issue is it? If re2 is available via the OS package manager on that platform, we could try to find and use that.

**Comment by tuncer** [comment]
Trying to detect and use the OS-provided re2 library can be done via a rebar.config.script that looks for the library (in well-known places) and modifies/generates  port_env accordingly. We could use autoconf, but that would complicate the build process.

**Comment by tuncer** [comment]
Work in progress: https://github.com/tuncer/re2/compare/rebar3

TODO: should we check an env var to ignore system re2?

@jj1bdx, can you verify if the config logic to detect clang with libc++ works on FreeBSD?

**Comment by tuncer** [comment]
To be clear, some of it, like system re2 detection, needs a little more work to be solid.

**Comment by tuncer** [comment]
@netDalek, does the rebar3 branch work for you? To be clear, it works with either rebar2 or rebar3, and the rebar3 support is preliminary, so for now stay with rebar2, until I figure out a couple details.

**Comment by tuncer** [comment]
Needs a little more work on BSD, but distinction between libstdc++ and libc++ works fine.

**Comment by tuncer** [comment]
https://github.com/tuncer/re2/compare/rebar3 works for me on linux-libstdc++ and freebsd-libc++, with or without system re2. @netDalek, does it work for you?

**Comment by netDalek** [comment]
I’m sorry for late answer. Now it works with current master of https://code.googlesource.com/re2 and gcc updated from 4.4 to 4.8. So now it works, but what about pin to exact https://code.googlesource.com/re2 version and not use master? It is quite frustrating to get unbuildable project when you use exact versions in your rebar.config

**Comment by tuncer** [comment]
It's good that it works with a newer gcc, but I've implemented default-linking against system-provided re2 lib (libre2.so), so I was curious to see if it works for you as well. This means, that we only fetch and build re2 if we cannot find it installed in the system.

As I wrote above, I've pinned it to the tag 2016-02-01 for the 2.2.1 release, but to implement the C++11 support I had to use master. I can pin it again, when the next upstream tag is created, but we cannot avoid C++11, unfortunately, unless they drop C++11, which is unlikely.

**Comment by tuncer** [comment]
Yesterday upstream has released 2016-03-01, and I've pinned that in c_src/build_deps.sh. I've also added the option to force master re2 by setting the env var RE2_MASTER.

**Comment by tuncer** [comment]
@cooldaemon, @janin, @djnym, @dhull, @matehat, would the planned change  from the rebar3 branch (linked above) to link against the system re2 lib (if available), and only fetch, build, and statically link a local build otherwise, in any way preclude you from using re2? The big change is that, unless you force a local re2 copy (via env var `NO_SYSTEM_RE2`), re2_nif.so will have a runtime dependency for libre2.so. If this is a problem, I can easily flip the logic to make it opt-in. I can imagine that deployment of the resulting re2_nif.so might fail due to a missing libre2.so in the target env.

**Comment by martin-langhoff** [comment]
(Working with @matehat) -- Reviewing and testing your 'Use system re2 if available' patch. From my perspective (with some background as Debian and Fedora packager), using system re2 is a sane default, and you have a clear override mechanism, so +1. 

If the default goes the other way, no problem at our end.

**Comment by tuncer** [comment]
The project started using a local copy at a time when distros didn't package re2, but my concern is that changing the default now will break some deployments. Therefore I'm wondering if we should make it opt-in, in order to not disrupt existing users' setups.

**Comment by martin-langhoff** [comment]
Current top of `rebar3` branch seems to DTRT at `make` time on Fedora 23, with `re2-devel` package installed. It does not fetch re2, it only builds `re2.erl` and `re2_nif.cpp`. 

However, my target platform is CentOS7/RHEL7 and there it fails the detection. The file listings of the rpms in F23 vs EL7 are the same, so finding the files / file paths should not be an issue. I could not spot in your patches a re2 minimum version requirement. Maybe EL7's `re2-20131024` is too old?

**Comment by tuncer** [comment]
No worries, I'm trying to get feedback, and I'm afraid the only response
I'll get is when the default is changed and users file tickets about broken
builds. I don't want to disrupt users' builds. Nobody downstream likes that
to happen.

​I'm also definitely aware of the advantages of relying on a system re2,
and I'd like to make it the default. Therefore, I'm trying to figure out
how to (1) inform every user beforehand, and/or (2) announce and flip the
default together with a major version bump.

Even if the default won't be changed in the next release, having a switch
will allow packagers to avoid the embedded re2.

**Comment by tuncer** [comment]
On 1 March 2016 at 19:01, Martin Langhoff <notifications@github.com> wrote:
>
> Current top of rebar3 branch seems to DTRT at make time on Fedora 23, with re2-devel package installed. It does not fetch re2, it only builds re2.erl and re2_nif.cpp.

DTRT?

How does re2_nif.cpp build, when the system re2 lib is incompatible?

> However, my target platform is CentOS7/RHEL7 and there it fails the detection. The file listings of the rpms in F23 vs EL7 are the same, so finding the files / file paths should not be an issue. I could not spot in your patches a re2 minimum version requirement. Maybe EL7's re2-20131024 is too old?

The only time I had adapt to an API change was in 2011
(1467f6904c79dba8e733de1359342f4795833119), so re2-20131024 shouldn't
be a problem.

Can you paste the file listing of both RPMs?

**Comment by martin-langhoff** [comment]
DTRT = Do The Right Thing. In other words, on Fedora 23 it's looking good.

Both EL7 and F23 have this listing for `re2` and `re2-devel`:
```
/usr/include/re2
/usr/include/re2/filtered_re2.h
/usr/include/re2/re2.h
/usr/include/re2/set.h
/usr/include/re2/stringpiece.h
/usr/include/re2/variadic_function.h
/usr/lib64/libre2.so
/usr/lib64/libre2.so.0
/usr/lib64/libre2.so.0.0.0
```

However, for some reason on EL7 the detection is not working. I'll be diving deeper into this tomorrow.

**Comment by tuncer** [comment]
@martin-langhoff, where did you find re2 packages for CentOS? I can't find anything in a normal repo for CentOS 6 or 7. What's the name of the RPM?

**Comment by martin-langhoff** [comment]
Update: Actually it works correctly on EL7. My mix.lock was outdated and resetting things to master resolved this. Apologies for the noise.

The re2 packages come from EPEL repo which are not part of RHEL/CentOS _base_, but are maintained by the Fedora/RH team and without EPEL RHEL/CentOS is rather limited, so everyone uses something from there...

**Comment by martin-langhoff** [comment]
On EL7 (CentOS7 in this case):
```
[vagrant@foo vagrant]$ sudo yumdownloader --urls re2 re2-devel
Loaded plugins: fastestmirror
Loading mirror speeds from cached hostfile
 * base: mirror.unl.edu
 * epel: mirrors.kernel.org
 * extras: mirror.oss.ou.edu
 * updates: centosv4.centos.org
http://mirrors.kernel.org/fedora-epel/7/x86_64/r/re2-20131024-1.el7.x86_64.rpm
http://mirrors.kernel.org/fedora-epel/7/x86_64/r/re2-devel-20131024-1.el7.x86_64.rpm

```
I don't think it is available on EL6, based on changelog...

**Comment by tuncer** [comment]
Great, it made no sense for detection to fail in that scenario.

**Comment by tuncer** [comment]
Merged the rebar3 branch to master after doing some minor changes, like reversing the system re2 check logic, and if you've been following the thread on the rebar list, you might be aware that there's some unresolved issues with rebar3 generally, but re2 now builds with either rebar2 or rebar3. Eventually we should look for a system re2 by default, but I couldn't enable it due to the high risk of disrupting existing users' environments because of the runtime requirement for libre2.so. Packagers can set the env var `SYSTEM_RE2` to try and use a system libre2.so.

Unfortunately, re2_nif.cpp also needs to be compiled with -std=c++11 now due to upstream changes, therefore to use re2 upstream 2016-03-01 or newer C++11 has to be enabled.

**Comment by tuncer** [comment]
Unless there's something I've missed, I'll close this ticket as resolved when I cut the next release. @netDalek?

**Comment by jj1bdx** [comment]
`make` and `rebar compile` worked on FreeBSD 10.3-PRERELEASE r295134 with the stock clang. `make eunit` generated no error. Sorry for getting late.

**Comment by tuncer** [comment]
Thanks @jj1bdx, as I wrote earlier, it also worked for me on freebsd-11.

**Comment by tuncer** [comment]
Fix included in 1.4.0.
